src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -0,0 +1,866 @@
+                }
+            }
+
+            if (segment == null)
+            {
+                // No free tail space, allocate a new segment
+                segment = CreateSegmentUnsynchronized();
+                segment.SetMemory(_pool.Rent(Math.Max(_minimumSegmentSize, count)));
+            }
+
+            if (_commitHead == null)
+            {
+                // No previous writes have occurred
+                _commitHead = segment;
+            }
+            else if (segment != _commitHead && _commitHead.Next == null)
+            {
+                // Append the segment to the commit head if writes have been committed
+                // and it isn't the same segment (unused tail space)
+                _commitHead.SetNext(segment);
+            }
+
+            // Set write head to assigned segment
+            _writingHead = segment;
+
+            return segment;
+        }
+
+        private BufferSegment CreateSegmentUnsynchronized()
+        {
+            if (_pooledSegmentCount > 0)
+            {
+                _pooledSegmentCount--;
+                return _bufferSegmentPool[_pooledSegmentCount];
+            }
+
+            return new BufferSegment();
+        }
+
+        private void ReturnSegmentUnsynchronized(BufferSegment segment)
+        {
+            if (_pooledSegmentCount < _bufferSegmentPool.Length)
+            {
+                _bufferSegmentPool[_pooledSegmentCount] = segment;
+                _pooledSegmentCount++;
+            }
+        }
+
+        internal void Commit()
+        {
+            // Changing commit head shared with Reader
+            lock (_sync)
+            {
+                CommitUnsynchronized();
+            }
+        }
+
+        internal void CommitUnsynchronized()
+        {
+            if (_writingHead == null)
+            {
+                // Nothing written to commit
+                return;
+            }
+
+            if (_readHead == null)
+            {
+                // Update the head to point to the head of the buffer.
+                // This happens if we called alloc(0) then write
+                _readHead = _commitHead;
+                _readHeadIndex = 0;
+            }
+
+            // Always move the commit head to the write head
+            _commitHead = _writingHead;
+            _commitHeadIndex = _writingHead.End;
+            _length += _currentWriteLength;
+
+            // Do not reset if reader is complete
+            if (_maximumSizeHigh > 0 &&
+                _length >= _maximumSizeHigh &&
+                !_readerCompletion.IsCompleted)
+            {
+                _writerAwaitable.Reset();
+            }
+
+            // Clear the writing state
+            _writingHead = null;
+            _currentWriteLength = 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void Advance(int bytesWritten)
+        {
+            if (_writingHead == null)
```

Comment:
When can writing head be null? Is the exception going to propagate to user code? If yes, why?

Reply:
If you call Advance without calling GetMemory()

Reply:
I assume the exception will explain the problem.

Reply:
The exceptions need to be reviewed. They are from before the IOutput refactor.